javalib/src/main/scala/java/util/Collections.scala

```diff
@@ -0,0 +1,1131 @@
+    class BasicCheckedList extends CheckedList[E] {
+      protected val elemClazz: Class[E] = typ
+      protected val inner: List[E] = list
+    }
+    list match {
+      case _: Serializable with RandomAccess =>
+        new BasicCheckedList with Serializable with RandomAccess
+      case _: Serializable => new BasicCheckedList with Serializable
+      case _: RandomAccess => new BasicCheckedList with RandomAccess
+      case _               => new BasicCheckedList
+    }
+  }
+
+  def checkedMap[K, V](m: Map[K, V], keyType: Class[K], valueType: Class[V]): Map[K, V] = {
+    class BasicCheckedMap extends CheckedMap[K, V, Map[K, V]] {
+      protected def keyClazz: Class[K] = keyType
+      protected def valueClazz: Class[V] = valueType
+      protected val inner: Map[K, V] = m
+    }
+    m match {
+      case _: Serializable => new BasicCheckedMap with Serializable
+      case _               => new BasicCheckedMap
+    }
+  }
+
+  def checkedSortedMap[K, V](m: SortedMap[K, V], keyType: Class[K], valueType: Class[V]): SortedMap[K, V] = {
+    class BasicCheckedSortedMap extends CheckedSortedMap[K, V] {
+      protected def keyClazz: Class[K] = keyType
+      protected def valueClazz: Class[V] = valueType
+      protected val inner: SortedMap[K, V] = m
+    }
+    m match {
+      case _: Serializable => new BasicCheckedSortedMap with Serializable
+      case _               => new BasicCheckedSortedMap
+    }
+  }
+
+  def emptyIterator[T](): Iterator[T] = {
+    new Iterator[T] {
+      override def hasNext: Boolean = false
+
+      override def next(): T =
+        throw new NoSuchElementException
+
+      override def remove(): Unit =
+        throw new IllegalStateException
+    }
+  }
+
+  def emptyListIterator[T](): ListIterator[T] = {
+    new ListIterator[T] {
+      override def hasNext: Boolean = false
+
+      override def next(): T =
+        throw new NoSuchElementException
+
+      override def hasPrevious: Boolean = false
+
+      override def previous(): T =
+        throw new NoSuchElementException
+
+      override def nextIndex(): Int = 0
+
+      override def previousIndex(): Int = -1
+
+      override def remove(): Unit =
+        throw new IllegalStateException
+
+      override def set(e: T): Unit =
+        throw new IllegalStateException
+
+      override def add(e: T): Unit =
+        throw new UnsupportedOperationException
+    }
+  }
+
+  def emptyEnumeration[T](): Enumeration[T] = {
+    new Enumeration[T] {
+      def hasMoreElements: Boolean = false
+
+      def nextElement(): T =
+        throw new NoSuchElementException
+    }
+  }
+
+  final def emptySet[T](): Set[T] =
+    EMPTY_SET.asInstanceOf[Set[T]]
+
+  final def emptyList[T](): List[T] =
+    EMPTY_LIST.asInstanceOf[List[T]]
+
+  final def emptyMap[K, V](): Map[K, V] =
+    EMPTY_MAP.asInstanceOf[Map[K, V]]
+
+  def singleton[T](o: T): Set[T] = {
+    unmodifiableSet(new AbstractSet[T] with io.Serializable {
```

Comment:
It is to make sure that `UnsupportedOperationException` is thrown directly if any of the methods that could modify the set are called. With `unmodifiableSet `:
```scala
val s = singleton("a")
s.remove("b")
```
will throw an exception on the `UnsupportedOperationException` with the `unmodifiableSet ` but without it it would just work fine because "b" is never found in the set and hence the removal attempt is never done. 
I grant that the specification is not that clear on what an "attempts to modify the returned set", but I believe that I would be better to fail fast when using any unsupported operation because that could only due to a bug is the user code.